### ISSUE TITLE
[RFC] ? and !

### ISSUE THREAD

**Issue by asterite** [crystal-lang/crystal#406]
Since `Nil` has a very important role in the language, we were thinking of changing the semantic of the language a bit so dealing with nils is easier and shorter.

The idea is to drop support for method names ending with "?" and "!". Instead, "?" would have the same meaning as the current [Object#try](http://crystal-lang.org/api/Object.html#try%28%26block%29-instance-method), while "!" would have the same meaning as [Object#not_nil!](http://crystal-lang.org/api/Object.html#not_nil%21-instance-method).

With this change in place we could have `hash[key]` always return a nilable value, and then you would write `hash[key]!` to get a non-nil value, raising if it's nil. So we wouldn't need to have a separate set of `#[]` and `#[]?` methods. For `Array#[]` I think the most reasonable thing would be to make it non-nilable with a possible raise of "index out of bounds", or we can just always stick to the convention and make `#[]` be nilable so you always know that `#[]` is nilable and `#[]!` might raise.

As a replacement for, for example `Array#empty?`, we can use `Array#is_empty`. For `Array#map` and `Array#map!` we can use `Array#map_in_place`, etc.

What do you think?

**Comment by jhass** [comment]
Sorry, but I don't like this at all.

The places where you actually use .not_nil! in application code seem to be rare, it's just writing the standard library were it might feel as noise. `is_empty` and `map_in_place` are terrible, verbose names.

Unfortunately I have nothing constructive to add, but I actually think that the current way isn't bad at all.

**Comment by Exilor** [comment]
I agree with @jhass.
Besides what he said, boolean methods ending with ? and "dangerous" methods ending with ! are a big part of what makes Ruby look like... well, Ruby. And since one of the stated goals of Crystal is to look as similar to Ruby as possible, I'd keep it that way. I think we already are in the perfect sweet spot of Ruby look and Crystal-specific functionality.

**Comment by vendethiel** [comment]
Let's huffmanize this and try to discourage people from using nil and using exceptions :-)

**Comment by brasten** [comment]
I like this idea very much. It's basically the way Swift deals with Optionals.

If that change is too drastic for people to take, something similar would be great.

Perhaps ! and ? starting the method call?  `session.user.?address.?line_one`? I'd prefer the original suggestion to that, though.

**Comment by akaufmann** [comment]
@Exilor 
Perhaps we come to better / different ideas than Ruby and if it fits we should not go the "Ruby Way" just because Ruby does it. Just my two cents.
As @brasten says, the "Swift Way" is nice. We should use it as well after types (prevents Object#not_nil! is spread all over the place).

```crystal
hash = {} of String => String!
i :: Int!
```

or like Nim
```nim
hash = {} of String => String not nil
``` 

Maybe we should start RFCs like Rust ;-)

**Comment by Exilor** [comment]
@akaufmann A big part of the appeal Crystal has is, to put it simply, "it's aaaalmost like Ruby but compiled!". The further it deviates from that, the more that appeal is lost.
It's a slippery slope. And once the language has deviated enough, why bother with yet another language that tries to reinvent the wheel? 
It's very easy to pick up Crystal if you know Ruby. It's a ridiculously small learning curve. If that changes I don't know why anyone would choose to invest time in learning it instead of playing it safe and learning a more popular language.

**Comment by waj** [comment]
@Exilor Crystal resembles Ruby because we like the design and philosophy of the language, but it doesn't pretend to be in any manner a drop in replacement. There are too many things already that are different and most of the differences come from the fact that Crystal is a compiled and strongly typed language. The treatment for the nilable variables is one of these differences and I believe one of the strongest points in Crystal. So for me it is reasonable that the primitives for handling these values are tight very deep in the language already.

**Comment by waj** [comment]
@akaufmann Sort of that feature is already present in Crystal:
```crystal
getter! foo
```
When you call `foo` it will raise if `@foo` is nil. Of course the difference is that in Swift can access the instance variable and get the assertion anyway.

**Comment by Exilor** [comment]
@waj It may not be intended to be a drop-in replacement but I firmly believe that whatever (modest) interest it has garnered so far is mostly thanks to its similarity to Ruby. Pardon the harshness, but in my opinion it's Crystal's greatest asset and without it there's not much else that makes it stand out.

**Comment by akaufmann** [comment]
@Exilor Don't get me wrong. I love Ruby as well Crystal. But I also love other languages (Rust, Swift, ...). They have also good ideas. I don't think a few difference here and there keeps People away. You can say the opposite as well: Why should Ruby Programmer care about Crystal? Just because it is faster (nah, there are great feature in crystal)? If it has some more great features, what ruby does not have, maybe they are more impressed. For example, I don't like Ruby's Object#freeze method. I hope that Crystal don't implement it. I hope **let** for immutable variables will come.

@waj I know, but not for local variables :-)

**Comment by asterite** [comment]
To give some more context about this discussion, when we started developing Crystal and decided that we wanted to explicitly deal with Nil, we knew that we would have to have a way to say "this is not nil, raise if so" and "do this only if it's not nil (otherwise do nothing)". But since "!" and "?" were already taken we used "try" and "not_nil!" as the initial placeholders.

Yesterday we were toying with Swift (not because of this), and we learned how they deal with nil. So @waj suggested that we could do the same. I told him that we *could maybe* do that, but that I wasn't entirely sure. But, we could ask the community to have more opinions and a better way to decide.

As for me, I wouldn't make the change for these reasons:

1. It's not used enough to justify a whole special syntax for it. In the whole Crystal repository (including src, samples and spec) I find 192 occurrences of `not_nil!` and 242 of `try`. The whole repository has 98060 lines of code right now. That's a %0.4. In [DeBot](https://github.com/jhass/DeBot) I found 18 `not_nil!` and 9 `try!` (in about 4000 lines). In [prax.cr](https://github.com/ysbaddaden/prax.cr) I found 1 `not_nil!` and no `try` (in about 1000 lines). Most of the `not_nil!` might probably be hidden behind a [property!](http://crystal-lang.org/api/Object.html#property%21%28%2Anames%29-macro), but that's perfectly fine: an idiom that covers this case, without the need for a special rule in the language. Other nilables might be hidden behind `if`, but that's again perfectly fine, and a nice way to deal with nil without an explicit syntax. And, as we can see (or at least the numbers say so), `nil` is not as pervasive as we might think.
2. It adds more rules to the language. Right now `try` and `not_nil!` are regular methods: they have documentation, you can see their implementation. If we want to change the message raised by `not_nil!` we can go to [this line](https://github.com/manastech/crystal/blob/master/src/nil.cr#L117) and change it. If it's a rule buried in the language it must be documented somewhere else, and nobody can change it (of course, we could make "!" and "?" rewrite to regular methods, but what would be the point then?). Granted, people will have to learn about `try` and `not_nil!` eventually, but they are just regular methods, not something special.
3. It breaks some of the Ruby feeling that the language has. If I write `if array.is_empty` it doesn't feel right for me. Maybe it's just because I'm used to that, but so are the many Ruby programmers that might stumble upon this language.
4. It would make things a bit inconsistent. Right now `#[]` raises and `#[]?` returns nil, **always**. By not having a `#[]?` method (just `#[]`) we would have to make it be nilable. But in Swift for arrays it raises, so that's an inconsistency. What if I want to get an element in an index or nil if it's not there? Maybe use another name, like "fetch" or "at", but... I don't like inconsistencies, as they always make you think twice.

In short (well, long), I think the benefits would be very small.

But I was surprised to find out that many actually think that Swift's way can actually be a good thing (that is, for Crystal). So @waj is definitely not alone here.

I guess we can continue discussing until we find more convincing reasons to do or not to do this feature.

@akaufmann I like the idea of RFCs. I think just creating an issue with `[RFC]` on the title is enough for now :-)

**Comment by asterite** [comment]
I actually have another reason :-)

Right now if you do `hash[key]` and the key is not found, you get a "Missing key in hash" error message, which is much more descrptive than "nil assertion failed". That is, one can choose how to implement `#[]` and `#[]?` (and other raise/nil method pairs) to provide better error messages, and also giving an opportunity to make the `?` method more efficient. For example [this method](https://github.com/manastech/crystal/blob/master/src/llvm/function_collection.cr#L17) does that, and it could also be improved by not [creating a Function object](https://github.com/manastech/crystal/blob/master/src/llvm/function_collection.cr#L24) (although in this case there's no difference because Function is a struct).

**Comment by ysbaddaden** [comment]
I'd be very sad to see #empty? and other #blank? or #any? methods go away. It may be my favorite little Ruby tweak :)

I also like the fact that Crystal is more harsh by raising when [] would be nil or is out of bounds. Your code failed because you expected something that wasn't there, instead of failing later because of a nil. I think I like []? too for checking existence, but I never used it yet.

**Comment by vendethiel** [comment]
This is a very interesting discussion, but one that seems to assume most crystal code will have to deal with lots of Nil. I think we should try to avoid that :-) (and indeed, @asterite's numbers seem to go that way).

I'm not "in love" with marks in method names, I'm just arguing against it because I'd like Nil to be almost-unexistant in most code bases :) (and so don't want to sacrifice 2 very useful punctuation symbols to it)

**Comment by ysbaddaden** [comment]
Yes, @vendethiel !

**Comment by asterite** [comment]
As a side note, both @waj and I agree that `not_nil!` is an ugly name, so any suggestion for a better name (if we finally retain it) is very welcome :smiley: 

We thought of `unrwap`, but that would be more appropriate for an Optional type. Here it's just a union of Nil and other types, so that name probably doesn't make sense.

**Comment by Marand-** [comment]
> It adds more rules to the language. Right now try and not_nil! are regular methods: they have documentation, you can see their implementation. 

This is probably the best reason not to turn `?` and `!` into special operators, in my opinion.  I don't love the Ruby-style `method?` and `method!` conventions, but I *do* like the design of having as few keywords as possible and representing everything as objects and methods, including operators.  

Also, it seems like disallowing `!` and `?` in method names would complicate things, considering `!` is a method that can be changed to redefine negation.  (Odd thing I noticed:  Crystal only allows calling it as `!x`, but in Ruby you can call it as `!x` or `x.!`.  Is that a bug in Crystal?)

>As a side note, both @waj and I agree that not_nil! is an ugly name, so any suggestion for a better name (if we finally retain it) is very welcome 

I can't think of anything better; it may be ugly but it's also pretty self-explanatory.  The only other thing I can think of would be to allow `?` as a method name and do `Object?` or `Object.?` instead of `Object.not_nil!`.  I think that would look better in code (especially the former), but I have a feeling it would be a bad idea because of things like the `? :` operator and maybe a lack of consistency with how `?` is used elsewhere.

**Comment by ysbaddaden** [comment]
I think `Object?` already says that this is a nilable union type: `Object | Nil`.

**Comment by Marand-** [comment]
> I think Object? already says that this is a nilable union type: Object | Nil.

Really?  Didn't see it in the Object API page anywhere, and I got syntax errors when I checked with a quick empty class test, so I thought it was unused.  Tested with this:
```
class A
end
a = A.new
````
Trying to use `A?` gave an `unexpected token: EOF` message, and `a?` gave `undefined local variable or method 'a?'`.

**Comment by asterite** [comment]
Types have a separate grammar, and that grammar is only available in a few places: type restrictions, generic type arguments and variable declaration. So for example:

```crystal
Array(String?).new # ok
String? # error

def foo(x : String?) # ok
end

Array(String*) # ok, pointer to string

def foo(x : String*) # ok
end
```

We could remove this and always use full names everywhere (for example you'd have to write Pointer(Int32) instead of Int32*), but I think that would make C bindings much harder to read and write. Also, `|` is a "bitwise or" operation, but in the type grammar is a union. So:

```crystal
String | Int32 #=> undefined method '|' for String:Class
Array(String | Int32) # ok
```

Again, this *could* be changed so that you'd have to write `Array(Union(String, Int32))`, so things are more consistent, but it will greatly affect readability, I believe. There's also the type grammar for `Int32 -> Int32`, which one can also write as `Proc(Int32, Int32)`. And also `{Int32, Char}`, which is `Tuple(Int32, Char)`.

**Comment by Marand-** [comment]
I see.  I didn't even think of testing it in types, my head's still in "everything is an object" so I just went looking for methods.  If `?` as a method isn't feasible because of that, then I'm at a loss for a better suggestion than `not_nil!` unfortunately.  Oh well.

**Comment by ozra** [comment]
As a complete non-Rubyist (@Exilor - Crystal is an excellent language, that's why _I_ like it (C++er)) I've come to really like this style of, _"duck-wording"_, especially `dangerous!(foo)` is a really smart practise I think. Like @vendethiel mentions: mutation is bad; but sometimes it's a performent necessary evil - and then it's better to have a convention for naming, so the danger is obvious at the call site. I'm very used to `foo.is-empty()`, `foo.isEmpty()` and similar wordings from other languages, but same here: must say I like the look of `foo.empty?`.

As for the `foo.not_nil!`, it's not impossible to at least shorten it `foo.nonil!` or _something_ to make it more 'keywordish/operatorish' (compare `elsif`). It's nice if all "language central" constructs can be free of underscores, or well, as terse as possible, and stay lean and mean. It gets mind share.

And the very important question @jhass already raised, which seems to have its answer in @asterite quick checks: there doesn't seem to be enough actual uses in practise to warrant special syntax.

But for Maybes and Optionals, the questions might be re-raised..

In type annotations `?` feels super natural for meaning `| Nil`, but that's another thing.

**Comment by sergey-kucher** [comment]
> As a side note, both @waj and I agree that not_nil! is an ugly name, so any suggestion for a better name (if we finally retain it) is very welcome

Relying on the fact *nil value means non-existing value*, I propose the next replacement for `#not_nil!`

- `#exists!`

`!` symbol is necessary because method can produce exception

**Comment by ktec** [comment]
perhaps the ActiveSupport idea of #present?

On 25 July 2015 at 11:13, Sergey Kucher <notifications@github.com> wrote:

> As a side note, both @waj <https://github.com/waj> and I agree that
> not_nil! is an ugly name, so any suggestion for a better name (if we
> finally retain it) is very welcome
>
> Relying on the fact *nil value means non-existing value*, I propose the
> next replacement for #not_nil!
>
>    - #exists!
>
> ! symbol is necessary because method can produce exception
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/manastech/crystal/issues/406#issuecomment-124831607>.
>



-- 
keithsalisbury@gmail.com

**Comment by ozra** [comment]
or  perhaps:
```crystal
if val.some?
    p val.some!  # instead of not_nil!
end
p "Has no value (nil)"  if val.none?   # similar to is_empty? (rather is_nil?)
```
"none" goes with the notion that nil is not really a value, but rather Nil is bottom type.

**Comment by kirbyfan64** [comment]
Ok...

I'm not really a Ruby user, but I love Crystal, and I have come to love the `?` and `!` characters for method names.

Also:

- If something can throw an exception (like `not_nil!`), I want it to be explicit. Not that short.
- The `?` suffix is great for boolean attributes.

**Comment by sergey-kucher** [comment]
+1 for `#some!` and `#none?`, instead of `#not_nil!` and `#nil?`

**Comment by technorama** [comment]
What about `.?`

Examples:
`foo.?.bar.baz` would be equivalent to `foo.try &.bar.baz`
`foo.?.bar.?.baz` would be equivalent to `foo.try &.bar.try &.baz`

Short circuit functionality is required, just like try.  If bar is nil in the first example, baz should never be attempted.

**Comment by ozra** [comment]
I like @technorama 's idea. Just a small twist, why not skip the dot after question mark:
```crystal
foo.?bar.baz
foo.?bar.?baz
```
Pretty much "reverse" of the common `a?.b?.c` pattern - but works for Crystal, since `?` can't be used as _initial_ identifier part.

**Comment by technorama** [comment]
@ozra It looks like part of the method name.   `bar?` is a name but `?bar` is something else entirely is hard to explain to users.

**Comment by ozra** [comment]
@technorama - yeah, that's a good point.

**Comment by Limeth** [comment]
Not a ruby user, but I like the `?` and `!` at the end of method names. Adding custom behavior for these symbols would make the code less readable.
As @sergey-kucher said,
> +1 for #some! and #none?, instead of #not_nil! and #nil?

**Comment by asterite** [comment]
I'm closing this. One problem @waj noted with Swift is that `!` ends the program abruptly. Now, we could do the same in Crystal, but maybe a better way, so users can redefine the meaning, is to make `!` invoke `not_nil!`. But if that's the case then it's probably better just to be explicit here and type `not_nil!`. Also, `!` is so easy to type you feel more tempted to use. `not_nil!` looks ugly, so you are actually tempted to avoid using it, which is good.

Then there's the conflict with current methods ending with `?`, so I'm closing this issue for now.
